iris-mpc-cpu/src/network/grpc.rs

```diff
@@ -43,76 +44,120 @@ impl MessageQueueStore {
+        Ok(())
     }
 
-    fn get_channel(&self, party_id: &Identity) -> eyre::Result<QueueChannel> {
-        let channel = self.queues.get(party_id).ok_or_eyre(format!(
-            "Channel not found for party {:?}, existing channels: alice {}, bob {}, charlie {}",
-            party_id,
-            self.queues.contains_key(&Identity("alice".into())),
-            self.queues.contains_key(&Identity("bob".into())),
-            self.queues.contains_key(&Identity("charlie".into()))
-        ))?;
-        Ok((*channel).clone())
+    async fn pop(&self, sender_id: &Identity) -> eyre::Result<Vec<u8>> {
+        let queue = self.queues.get(sender_id).ok_or(eyre!(format!(
+            "RECEIVE: Sender {sender_id:?} hasn't been found in the message queues"
+        )))?;
+
+        let mut queue = queue.lock().await;
+
+        let msg = queue.next().await.ok_or(eyre!("No message received"))??;
+
+        Ok(msg.data)
     }
+}
+
+struct OutgoingStream {
+    streams: HashMap<SessionId, Arc<UnboundedSender<SendRequest>>>,
+}
 
-    pub async fn push_back(&self, party_id: &Identity, value: Vec<u8>) -> eyre::Result<()> {
-        let channel = self.get_channel(party_id)?;
-        // sends the value via the channel sender; if failed, returns an error
-        channel.sender.send(value).await.map_err(|e| e.into())
+impl OutgoingStream {
+    fn new() -> Self {
+        OutgoingStream {
+            streams: HashMap::new(),
+        }
     }
 
-    pub async fn pop_front(&self, party_id: &Identity) -> eyre::Result<Vec<u8>> {
-        let channel = self.get_channel(party_id)?;
-        channel.receiver.recv().await.map_err(|e| e.into())
+    fn add_session_stream(&mut self, session_id: SessionId, stream: UnboundedSender<SendRequest>) {
+        self.streams.insert(session_id, Arc::new(stream));
+    }
+
+    fn get_stream(
+        &self,
+        session_id: &SessionId,
+    ) -> eyre::Result<Arc<UnboundedSender<SendRequest>>> {
+        self.streams
+            .get(session_id)
+            .cloned()
+            .ok_or(eyre!("Streams for Session {session_id:?} not found"))
     }
 }
 
+#[derive(Default, Clone)]
+pub struct GrpcConfig {
+    pub timeout_duration: Duration,
+}
+
+// WARNING: this implementation assumes that messages for a specific player
+// within one session are sent in order and consecutively. Don't send messages
+// to the same player in parallel within the same session. Use batching instead.
 #[derive(Clone)]
 pub struct GrpcNetworking {
-    party_id:       Identity,
+    party_id:         Identity,
     // other party id -> client to call that party
-    clients:        Arc<DashMap<Identity, PartyNodeClient<Channel>>>,
-    message_queues: Arc<DashMap<SessionId, MessageQueueStore>>,
+    clients:          Arc<Mutex<HashMap<Identity, PartyNodeClient<Channel>>>>,
+    // other party id -> outgoing streams to send messages to that party in different sessions
+    outgoing_streams: Arc<DashMap<Identity, OutgoingStream>>,
```

Comment:
why is this identity -> session_id -> stream and incoming is session_id -> identity -> stream?
I think this should be unified?
also maybe it would be good to make this 1 jump only, by just using a key that combines both? contains the party as well?

Reply:
Yep, good point. Will try to fix it.

Reply:
Ok, I managed to simplify the dashmap of OutgoingStream, but for incoming ones I have to have the current setup to correctly initialize them before any message is sent.